File: ios/Shared/Sources/Features/LmsRutimeTableParser.swift
import Foundation
import MoodleNativeNetworking

public enum LmsRutimeTableParser {
    public struct CoursePlacement: Equatable, Sendable {
        public let courseID: Int
        public let title: String
        public let room: String?
        public let dayIndex: Int?
        public let periodIndex: Int?
        public let detailURL: URL?
        public let isFavorite: Bool
        public let hasUnreadAnnouncement: Bool
        public let hasPendingAssignment: Bool
        public let hasUnreadForum: Bool
    }

    private static let dayMap: [Character: Int] = [
        "月": 0,
        "火": 1,
        "水": 2,
        "木": 3,
        "金": 4,
    ]

    public static func parse(from snapshot: LmsWebServiceClient.DashboardSnapshot) -> [CoursePlacement] {
        parse(dashboardBlocks: snapshot.dashboardBlocks, siteURL: snapshot.siteURL)
    }

    public static func parse(
        dashboardBlocks: [LmsWebServiceClient.DashboardBlock],
        siteURL: String
    ) -> [CoursePlacement] {
        let block = dashboardBlocks.first { block in
            if block.name == "rutime_table" {
                return true
            }

            guard let content = block.contents?.content else {
                return false
            }

            return content.contains("class=\"subject\"")
                && content.contains("active-course-name")
        }

        guard let content = block?.contents?.content, content.isEmpty == false else {
            return []
        }

        return parse(content: content, siteURL: siteURL)
    }

    public static func parse(content: String, siteURL: String) -> [CoursePlacement] {
        let normalized = content.replacingOccurrences(of: "\r\n", with: "\n")

        let tablePlacements = parseTablePlacements(content: normalized, siteURL: siteURL)
        if tablePlacements.isEmpty == false {
            return tablePlacements
        }

        let subjects = captureGroups(
            pattern: #"<div\s+class=\"subject\">([\s\S]*?)(?=<div\s+class=\"subject\">|</table>|$)"#,
            in: normalized
        )

        var placements: [CoursePlacement] = []

        for subject in subjects {
            placements.append(contentsOf: parseSubjectPlacements(subject, siteURL: siteURL))
        }

        return placements
    }

    private static func parseTablePlacements(content: String, siteURL: String) -> [CoursePlacement] {
        let rows = captureGroups(pattern: #"<tr[^>]*>([\s\S]*?)</tr>"#, in: content)
        var placements: [CoursePlacement] = []

        for row in rows {
            guard
                let periodText = firstMatch(pattern: #"<td\s+class=\"time\"[^>]*>([\s\S]*?)</td>"#, in: row),
                let periodNumber = Int(normalizedDigits(decodeHTMLText(periodText).trimmingCharacters(in: .whitespacesAndNewlines)))
            else {
                continue
            }

            let cells = captureGroups(
                pattern: #"<td\s+class=\"(?:highlight|empty)[^\"]*\"[^>]*>([\s\S]*?)</td>"#,
                in: row
            )

            for (cellIndex, cell) in cells.enumerated() {
                let subjects = captureGroups(
                    pattern: #"<div\s+class=\"subject\">([\s\S]*?)(?=<div\s+class=\"subject\">|$)"#,
                    in: cell
                )

                for subject in subjects {
                    placements.append(
                        contentsOf: parseSubjectPlacements(
                            subject,
                            siteURL: siteURL,
                            dayIndexOverride: cellIndex < 5 ? cellIndex : nil,
                            periodIndexOverride: periodNumber - 1
                        )
                    )
                }
            }
        }

        return placements
    }

    private static func parseSubjectPlacements(
        _ subject: String,
        siteURL: String,
        dayIndexOverride: Int? = nil,
        periodIndexOverride: Int? = nil
    ) -> [CoursePlacement] {
        guard
            let courseIDString = firstMatch(
                pattern: #"course/view\.php\?id=(\d+)"#,
                in: subject
            ),
            let courseID = Int(courseIDString)
        else {
            return []
        }

        let title = decodeHTMLText(
            firstMatch(
                pattern: #"<a[^>]*class=\"active-course-name\"[^>]*>([\s\S]*?)</a>"#,
                in: subject
            ) ?? ""
        )
        .trimmingCharacters(in: .whitespacesAndNewlines)

        let roomText = decodeHTMLText(
            firstMatch(
                pattern: #"<div\s+class=\"room\">([\s\S]*?)</div>"#,
                in: subject
            ) ?? ""
        )
        .trimmingCharacters(in: .whitespacesAndNewlines)

        let (parsedDayIndex, parsedPeriodIndices, room) = parseRoom(roomText)
        let detailURL: URL?
        if let href = firstMatch(pattern: #"<a[^>]*href=\"([^\"]+)\""#, in: subject) {
            detailURL = resolvedURL(from: href, siteURL: siteURL)
        } else {
            detailURL = URL(string: "\(siteURL)/course/view.php?id=\(courseID)")
        }

        let dayIndex = dayIndexOverride ?? parsedDayIndex
        let periodIndices = if let periodIndexOverride {
            [periodIndexOverride]
        } else {
            parsedPeriodIndices
        }

        if periodIndices.isEmpty {
            return [
                CoursePlacement(
                    courseID: courseID,
                    title: title,
                    room: room,
                    dayIndex: dayIndex,
                    periodIndex: nil,
                    detailURL: detailURL,
                    isFavorite: isIconOn("favouriteicon", in: subject),
                    hasUnreadAnnouncement: isIconOn("newsicon", in: subject),
                    hasPendingAssignment: isIconOn("assignicon", in: subject),
                    hasUnreadForum: isIconOn("forumicon", in: subject)
                )
            ]
        }

        return periodIndices.map { periodIndex in
            CoursePlacement(
                courseID: courseID,
                title: title,
                room: room,
                dayIndex: dayIndex,
                periodIndex: periodIndex,
                detailURL: detailURL,
                isFavorite: isIconOn("favouriteicon", in: subject),
                hasUnreadAnnouncement: isIconOn("newsicon", in: subject),
                hasPendingAssignment: isIconOn("assignicon", in: subject),
                hasUnreadForum: isIconOn("forumicon", in: subject)
            )
        }
    }

    private static func parseRoom(_ value: String) -> (Int?, [Int], String?) {
        guard value.isEmpty == false else {
            return (nil, [], nil)
        }

        let parts = splitScheduleAndRoom(value)
        let scheduleText = parts.schedule
        let pattern = #"([月火水木金])"#
        guard let dayToken = firstMatch(pattern: pattern, in: scheduleText, group: 1),
            let dayCharacter = dayToken.first,
            let dayIndex = dayMap[dayCharacter]
        else {
            return (nil, [], value)
        }

        let room = parts.room?
            .trimmingCharacters(in: .whitespacesAndNewlines)

        return (
            dayIndex,
            periodNumbers(in: scheduleText).map { $0 - 1 },
            room?.isEmpty == true ? nil : room
        )
    }

    private static func splitScheduleAndRoom(_ value: String) -> (schedule: String, room: String?) {
        guard let separatorIndex = value.firstIndex(where: { $0 == ":" || $0 == "：" }) else {
            return (value, nil)
        }

        let schedule = String(value[..<separatorIndex])
        let roomStartIndex = value.index(after: separatorIndex)
        let room = String(value[roomStartIndex...])
            .trimmingCharacters(in: .whitespacesAndNewlines)

        return (schedule, room)
    }

    private static func periodNumbers(in text: String) -> [Int] {
        guard let periodText = firstMatch(
            pattern: #"([月火水木金])\s*([0-9０-９]+(?:\s*[-－〜~～・･,，/／、]\s*[月火水木金]?\s*[0-9０-９]+)*)"#,
            in: text,
            group: 2
        ) else {
            return []
        }

        return captureGroups(pattern: #"([0-9０-９]+)"#, in: periodText)
            .compactMap { Int(normalizedDigits($0)) }
            .filter { $0 > 0 }
    }

    private static func normalizedDigits(_ value: String) -> String {
        value.unicodeScalars.map { scalar in
            let value = scalar.value
            if (0xFF10...0xFF19).contains(value),
                let scalar = UnicodeScalar(value - 0xFF10 + 0x30)
            {
                return String(scalar)
            }

            return String(scalar)
        }
        .joined()
    }

    private static func isIconOn(_ iconClass: String, in text: String) -> Bool {
        let escapedIconClass = NSRegularExpression.escapedPattern(for: iconClass)
        let pattern =
            #"(<span\s+class=\"on\">(?:(?!</span>)[\s\S])*?class=\"[^\"]*\b"# + escapedIconClass
            + #"\b[^\"]*\")"#
        return firstMatch(pattern: pattern, in: text) != nil
    }

    private static func decodeHTMLText(_ value: String) -> String {
        value
            .replacingOccurrences(of: "<[^>]+>", with: " ", options: .regularExpression)
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
    }

    private static func resolvedURL(from href: String, siteURL: String) -> URL? {
        if let absoluteURL = URL(string: href), absoluteURL.scheme != nil {
            return absoluteURL
        }

        guard let baseURL = URL(string: siteURL) else {
            return nil
        }

        return URL(string: href, relativeTo: baseURL)?.absoluteURL
    }

    private static func captureGroups(pattern: String, in text: String) -> [String] {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
            return []
        }

        let nsText = text as NSString
        let range = NSRange(location: 0, length: nsText.length)

        return regex.matches(in: text, options: [], range: range).compactMap { match in
            guard match.numberOfRanges > 1 else {
                return nil
            }

            let groupRange = match.range(at: 1)
            guard groupRange.location != NSNotFound else {
                return nil
            }

            return nsText.substring(with: groupRange)
        }
    }

    private static func firstMatch(pattern: String, in text: String, group: Int = 1) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
            return nil
        }

        let nsText = text as NSString
        let range = NSRange(location: 0, length: nsText.length)

        guard let match = regex.firstMatch(in: text, options: [], range: range),
            match.numberOfRanges > group
        else {
            return nil
        }

        let groupRange = match.range(at: group)
        guard groupRange.location != NSNotFound else {
            return nil
        }

        return nsText.substring(with: groupRange)
    }
}

File: android/features/src/main/kotlin/dev/example/moodlenative/features/LmsRutimeTableParser.kt
package dev.example.moodlenative.features

import java.net.URI

data class CoursePlacement(
    val courseID: Int,
    val title: String,
    val room: String?,
    val dayIndex: Int?,
    val periodIndex: Int?,
    val detailURL: URI?,
    val isFavorite: Boolean,
    val hasUnreadAnnouncement: Boolean,
    val hasPendingAssignment: Boolean,
    val hasUnreadForum: Boolean,
)

object LmsRutimeTableParser {
    private val dayMap = mapOf(
        '月' to 0,
        '火' to 1,
        '水' to 2,
        '木' to 3,
        '金' to 4,
    )

    fun parse(snapshot: LmsDashboardSnapshot): List<CoursePlacement> =
        parse(dashboardBlocks = snapshot.dashboardBlocks, siteURL = snapshot.siteURL)

    fun parse(dashboardBlocks: List<LmsDashboardBlock>, siteURL: String): List<CoursePlacement> {
        val block = dashboardBlocks.firstOrNull { block ->
            if (block.name == "rutime_table") {
                return@firstOrNull true
            }

            val content = block.contents?.content ?: return@firstOrNull false
            content.contains("class=\"subject\"") && content.contains("active-course-name")
        }

        val content = block?.contents?.content
        if (content.isNullOrEmpty()) {
            return emptyList()
        }

        return parse(content = content, siteURL = siteURL)
    }

    fun parse(content: String, siteURL: String): List<CoursePlacement> {
        val normalized = content.replace("\r\n", "\n")

        val tablePlacements = parseTablePlacements(normalized, siteURL)
        if (tablePlacements.isNotEmpty()) {
            return tablePlacements
        }

        val subjects = captureGroups(
            pattern = """<div\s+class="subject">([\s\S]*?)(?=<div\s+class="subject">|</table>|$)""",
            text = normalized,
        )

        return subjects.flatMap { parseSubjectPlacements(it, siteURL) }
    }

    private fun parseTablePlacements(content: String, siteURL: String): List<CoursePlacement> {
        val rows = captureGroups("""<tr[^>]*>([\s\S]*?)</tr>""", content)
        val placements = mutableListOf<CoursePlacement>()

        for (row in rows) {
            val periodText = firstMatch("""<td\s+class="time"[^>]*>([\s\S]*?)</td>""", row)
                ?.let { decodeHTMLText(it).trim() }
            val periodNumber = periodText
                ?.let(::normalizedDigits)
                ?.toIntOrNull()
                ?: continue

            val cells = captureGroups(
                pattern = """<td\s+class="(?:highlight|empty)[^"]*"[^>]*>([\s\S]*?)</td>""",
                text = row,
            )

            for ((cellIndex, cell) in cells.withIndex()) {
                val subjects = captureGroups(
                    pattern = """<div\s+class="subject">([\s\S]*?)(?=<div\s+class="subject">|$)""",
                    text = cell,
                )

                for (subject in subjects) {
                    placements += parseSubjectPlacements(
                        subject = subject,
                        siteURL = siteURL,
                        dayIndexOverride = cellIndex.takeIf { it < 5 },
                        periodIndexOverride = periodNumber - 1,
                    )
                }
            }
        }

        return placements
    }

    private fun parseSubjectPlacements(
        subject: String,
        siteURL: String,
        dayIndexOverride: Int? = null,
        periodIndexOverride: Int? = null,
    ): List<CoursePlacement> {
        val courseID = firstMatch("""course/view\.php\?id=(\d+)""", subject)
            ?.toIntOrNull()
            ?: return emptyList()

        val title = decodeHTMLText(
            firstMatch("""<a[^>]*class="active-course-name"[^>]*>([\s\S]*?)</a>""", subject) ?: "",
        ).trim()

        val roomText = decodeHTMLText(
            firstMatch("""<div\s+class="room">([\s\S]*?)</div>""", subject) ?: "",
        ).trim()

        val parsedRoom = parseRoom(roomText)
        val detailURL = firstMatch("<a[^>]*href=\"([^\"]+)\"", subject)
            ?.let { resolvedURL(it, siteURL) }
            ?: resolvedURL("$siteURL/course/view.php?id=$courseID", siteURL)

        val dayIndex = dayIndexOverride ?: parsedRoom.dayIndex
        val periodIndices = periodIndexOverride?.let(::listOf) ?: parsedRoom.periodIndices

        if (periodIndices.isEmpty()) {
            return listOf(
                CoursePlacement(
                    courseID = courseID,
                    title = title,
                    room = parsedRoom.room,
                    dayIndex = dayIndex,
                    periodIndex = null,
                    detailURL = detailURL,
                    isFavorite = isIconOn("favouriteicon", subject),
                    hasUnreadAnnouncement = isIconOn("newsicon", subject),
                    hasPendingAssignment = isIconOn("assignicon", subject),
                    hasUnreadForum = isIconOn("forumicon", subject),
                ),
            )
        }

        return periodIndices.map { periodIndex ->
            CoursePlacement(
                courseID = courseID,
                title = title,
                room = parsedRoom.room,
                dayIndex = dayIndex,
                periodIndex = periodIndex,
                detailURL = detailURL,
                isFavorite = isIconOn("favouriteicon", subject),
                hasUnreadAnnouncement = isIconOn("newsicon", subject),
                hasPendingAssignment = isIconOn("assignicon", subject),
                hasUnreadForum = isIconOn("forumicon", subject),
            )
        }
    }

    private fun parseRoom(value: String): ParsedRoom {
        if (value.isEmpty()) {
            return ParsedRoom(dayIndex = null, periodIndices = emptyList(), room = null)
        }

        val parts = splitScheduleAndRoom(value)
        val dayToken = firstMatch("""([月火水木金])""", parts.schedule)
        val dayIndex = dayToken?.firstOrNull()?.let(dayMap::get)

        if (dayIndex == null) {
            return ParsedRoom(dayIndex = null, periodIndices = emptyList(), room = value)
        }

        val room = parts.room?.trim()?.takeIf { it.isNotEmpty() }

        return ParsedRoom(
            dayIndex = dayIndex,
            periodIndices = periodNumbers(parts.schedule).map { it - 1 },
            room = room,
        )
    }

    private fun splitScheduleAndRoom(value: String): ScheduleAndRoom {
        val separatorIndex = value.indexOfFirst { it == ':' || it == '：' }
        if (separatorIndex < 0) {
            return ScheduleAndRoom(schedule = value, room = null)
        }

        return ScheduleAndRoom(
            schedule = value.substring(0, separatorIndex),
            room = value.substring(separatorIndex + 1).trim(),
        )
    }

    private fun periodNumbers(text: String): List<Int> {
        val periodText = firstMatch(
            pattern = """([月火水木金])\s*([0-9０-９]+(?:\s*[-－〜~～・･,，/／、]\s*[月火水木金]?\s*[0-9０-９]+)*)""",
            text = text,
            group = 2,
        ) ?: return emptyList()

        return captureGroups("""([0-9０-９]+)""", periodText)
            .mapNotNull { normalizedDigits(it).toIntOrNull() }
            .filter { it > 0 }
    }

    private fun normalizedDigits(value: String): String =
        buildString {
            for (char in value) {
                if (char.code in fullWidthZeroCode..fullWidthNineCode) {
                    append((char.code - fullWidthZeroCode + asciiZeroCode).toChar())
                } else {
                    append(char)
                }
            }
        }

    private fun isIconOn(iconClass: String, text: String): Boolean {
        val escapedIconClass = Regex.escape(iconClass)
        val pattern =
            """(<span\s+class="on">(?:(?!</span>)[\s\S])*?class="[^"]*\b$escapedIconClass\b[^"]*")"""
        return firstMatch(pattern, text, group = 1) != null
    }

    private fun decodeHTMLText(value: String): String =
        value
            .replace(Regex("<[^>]+>"), " ")
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")

    private fun resolvedURL(href: String, siteURL: String): URI? {
        val absoluteURL = runCatching { URI(href) }.getOrNull()
        if (absoluteURL?.isAbsolute == true) {
            return absoluteURL
        }

        val baseURL = runCatching { URI(siteURL) }.getOrNull() ?: return null
        return runCatching { baseURL.resolve(href) }.getOrNull()
    }

    private fun captureGroups(pattern: String, text: String): List<String> =
        Regex(pattern).findAll(text).mapNotNull { match ->
            match.groups[1]?.value
        }.toList()

    private fun firstMatch(pattern: String, text: String, group: Int = 1): String? =
        Regex(pattern).find(text)?.groups?.get(group)?.value

    private data class ParsedRoom(
        val dayIndex: Int?,
        val periodIndices: List<Int>,
        val room: String?,
    )

    private data class ScheduleAndRoom(
        val schedule: String,
        val room: String?,
    )

    private const val fullWidthZeroCode = 0xFF10
    private const val fullWidthNineCode = 0xFF19
    private const val asciiZeroCode = 0x30
}
